Introduce picklable configurables

gmrukwa opened the pull request:
#### What is the subject of this PR?

I was trying to implement a mechanism that allows the external configurables and scoped configurables get pickled.

Resolves #31 

#### Details

Since the only thing that we need is the decorated `__new__` or `__init__`, I thought that we could replace the instance that is actually returned to the instance of original class. This instance gets initialized with the decorated function. Through this we get:

- picklability of all configurable versions of picklable types (regardless it is scoped, external or plain configurable)
- more natural and intuitive inheritance: `isinstance(subclass_instance, type(superclass_instance))` evaluates to `True` even if superclass is an external configurable

#### Remarks

✅ I signed the CLA.

gmrukwa commented:
Hey @dhr , @sguada, could you please provide some feedback on this PR? If there's something I need to change before this gets merged, I would need to know. For now I have a workaround in my own code (thanks Python!), but I wanted to share this with others, as the issue seems quite frequent.

gmrukwa commented:
Hello @dhr, @sguada! Any progress?

gmrukwa commented:
Hello, is there anyone capable of reviewing?

pseeth commented:
I also just ran into this problem. A fix would be nice. Right now, the only thing I can do is wrap everything in functions to try and fix it so they get evaluated within the script and return the original class rather than the gin class.

gmrukwa commented:
@pseeth, as you see I am waiting for more than a month for the review. 😞

gmrukwa commented:
@pseeth you can make yourself a module ([like this one I did](https://github.com/gmrukwa/divik/blob/master/divik/core/_gin_bugfix.py)) and make sure it gets imported, so the gin works properly.

pseeth commented:
That looks nice! Thanks! I ended up writing an "unginifiy" configurable function which looks for the same class using  a `globals()` hack and then finds the associated arguments in the gin-config and reinstantiates the class: https://github.com/nussl/models/blob/master/src/gin_utils/__init__.py#L36

gmrukwa commented:
@dhr, @sguada, any progress?

sguada commented on gin/config.py:
```diff
@@ -267,6 +267,16 @@ class DecoratedClass(fn_or_cls):
+  elif subclass:
+    # this ensures picklability, as we produce the original class instance,
+    # the implicit inheritance is skipped
+    def hijacking_new(klass, *args, **kwargs):
```
hijacking_new -> _new

sguada commented on gin/config.py:
```diff
@@ -267,6 +267,16 @@ class DecoratedClass(fn_or_cls):
+      obj = fn_or_cls.__new__(klass, *args, **kwargs)
+      decorated_fn(obj, *args, **kwargs)
+      return obj
+    setattr(cls, '__new__', hijacking_new)
```
hijacking_new -> _new

sguada commented on tests/config_test.py:
```diff
@@ -295,6 +296,16 @@ def __init__(self, kwarg1=None, kwarg2=None):
+    self.kwarg2 = kwarg2
+
+config.external_configurable(
+  ExternalImportableClass, 'ExternalImportableClass')
```
add 2 extra spaces

sguada commented on tests/config_test.py:
```diff
@@ -641,6 +652,10 @@ def testConfigurableClass(self):
     self.assertEqual(instance.kwarg2, 'waldorf')
+    try:
+      pickle.dumps(instance)
+    except:
```
Be explicit about the type of exception you want to catch.

sguada commented on tests/config_test.py:
```diff
@@ -682,14 +697,22 @@ def testConfigurableSubclass(self):
 
+    try:
+      pickle.dumps(sub_instance)
+    except:
```
Be explicit about the type of exception you want to catch.

sguada commented on tests/config_test.py:
```diff
@@ -720,6 +743,17 @@ def testExternalConfigurableClass(self):
+    instance = configurable_class.kwarg1()
+    try:
+      pickle.dumps(instance)
+    except:
```
Be explicit about the type of exception you want to catch.

sguada commented on tests/config_test.py:
```diff
@@ -295,6 +296,16 @@ def __init__(self, kwarg1=None, kwarg2=None):
 
 
+class ExternalImportableClass(object):
+  """As above, but now importable"""
```
"""External Importable Class."""

sguada commented:
@gmrukwa sorry @dhr has found some issues this current solution for metaclasses. He will be adding more details here

gmrukwa commented:
> @gmrukwa sorry @dhr has found some issues this current solution for metaclasses. He will be adding more details here

I'd love to discuss it further 🙂 If I can do anything more to keep things going, just let me know.

gmrukwa commented:
Basically this version makes inheritance of a nested class fully transparent (like nonexistent), just influencing the parameters for object creation.

dhr commented on gin/config.py:
```diff
@@ -270,13 +270,13 @@ class DecoratedClass(fn_or_cls):
       if klass is DecoratedClass:
         klass = fn_or_cls
       obj = fn_or_cls.__new__(klass, *args, **kwargs)
       decorated_fn(obj, *args, **kwargs)
```
Thanks for the change! Initially I thought this would work, but after testing it internally a couple of issues came up. Specifically:

1) Calling decorated_fn here can break scoping. In order to apply config scopes to references (as in "@some_scope_name/ConfigurableClass()"), the configurable version of the function gets wrapped by yet another function that applies the scope. This happens when constructing a ConfigurableReference instance here, and reuses all the same decoration mechanisms (including subclassing for classes) as regular gin.configurable decoration does:

  https://github.com/google/gin-config/blob/484d03ab6f52c05455adb06541e9d8d0189d6967/gin/config.py#L301

  So for scoped configurable references, there are two layers of wrapping, one to apply the scope, and one to supply the configured parameters. Unfortunately, for that outer layer of wrapping, calling __new__ first before calling decorated_fn means that __new__ is called before the config_scope is opened (since that happens inside decorated_fn). That inner __new__ call then in turn calls its own decorated_fn (which is responsible for supplying parameters), but since it's not in the right scope the wrong parameters can be supplied. Hopefully that's somewhat clear...

2) The second issue is that this breaks for classes that use metaclasses, when the metaclass overrides __call__ and calls __init__ internally. One fairly broadly used example of this is in Sonnet v2's ModuleMetaclass here

 
 https://github.com/deepmind/sonnet/blob/5a465696383f967d5bffb6599347d9e6c15cef4b/sonnet/src/base.py#L126

  Here's a simple test case that fails for that reason (requires Sonnet):

  ```
import sonnet.v2 as snt

configurable_linear = gin.external_configurable(snt.Linear)

...

def testConfigurableSonnetModule(self):
    config_str = """
      Linear.output_size = 100
    """
    gin.parse_config(config_str)
    linear = configurable_linear()
    self.assertEqual(linear.output_size, 100)
```

Unfortunately I think those issues are hard to fix using the current approach... I haven't yet had a chance to think through alternative implementations in any depth, so not sure what other approaches might be feasible. Do you have any thoughts on another way to address this? (One approach that I'd like to get to is to replace a lot of this quite complicated ad-hoc decoration logic with wrapt, which would certainly change a lot of things implementation wise, but it's not clear to me it would fix the pickling issue specifically.)

gmrukwa replied:
Thanks for the review, I'll take a look into the points you mentioned.

The first one I didn't anticipate - the tests didn't fail, so I supposed it's all okay. I'll try to address that with a test case in the meantime.

The second seems more like a trade-off, as this is much more of what's doable with the configurable class. I know the priority is to keep the old stuff working, but an expectation to pickle the result of computations is quite standard I believe. I will think about keeping both for sure, but it's not obvious that there's a space for that.

Basically to ensure picklability, a class definition must be "static" (as static as for Python it could be) - it must be importable, it cannot be defined within a function. I'll review [the more complicated approach](https://stackoverflow.com/a/4648949/3067956) I found, but I will need to re-think how that's doable - an importable class must be accessible at the module scope anyway.

dhr replied:
Yeah it's unfortunate the first point isn't covered by the current tests... scoping is generally well tested but apparently scopes + references to external configurables seems to have been an overlooked case. A contribution there would be very welcome!

For the second point, since Sonnet is kind of an important use case I think it's necessary to maintain compatibility there. That said, an approach using __reduce(_ex)__ seems like it could be a good direction, and I'd think is more likely to be compatible with the current setup. Let me know if you look into it more and can tell if it would be feasible. Thanks again for working on this!

gmrukwa replied:
@dhr, based on [this](https://stackoverflow.com/a/6966942/3067956) and [this](https://stackoverflow.com/a/6798042/3067956) I think that the proper way to solve the second case is to actually wrap the `klass.__metaclass__.__call__` function directly, as that's what I was (indirectly) trying to achieve with my change. It would be a generalization of the current version of this fix for any metaclass and would probably address the first issue as well.

gmrukwa replied:
Also, right now you don't address the metaclasses in any specific way - you patch `__new__` or `__init__` only ([link](https://github.com/google/gin-config/blob/master/gin/config.py#L192-L198)). I'll prepare a test that replicates the Sonnet scenario you pointed out.

gmrukwa replied:
I've replicated the Sonnet issue in tests.

The solution with `klass.__metaclass__.__call__` will however slightly change the legacy behavior. `klass.__metaclass__.__call__` parameters will get patched, instead of `klass.__init__`.

I've found out that there was a difference between parameters going to `klass.__new__` and `klass.__init__`, which actually should be an issue. If your `klass.__new__` was supposed to return another class (let's say `int`), but had `klass.__init__` defined, it would return `int` only when parameters are passed in the code, not in the config. Here's test case describing the current behavior:

```python
@config.configurable
class ConfigurableClassWithNew(object):
  """A configurable class with __new__."""

  def __new__(cls, return_none=False):
    if return_none: # this will test __new__ and __init__ separation
      return None
    return super(ConfigurableClassWithNew, cls).__new__(cls)

  def __init__(self, return_none=False):
    pass

...

  def testConfigurableClassWithNewCanReturnOtherType(self):
    config_str = """
      ConfigurableClassWithNew.return_none = True
    """
    config.parse_config(config_str)
    instance = ConfigurableClassWithNew()
    self.assertIsNotNone(instance)

    config_str = ""
    config.parse_config(config_str)
    instance = ConfigurableClassWithNew(return_none=True)
    self.assertIsNone(instance)
```

After patching `klass.__metaclass__.__call__` in both cases we will have `None` (more what I'd expect). @dhr, @sguada will that change break antyhing for you?

gmrukwa replied:
Works with `__reduce__` 🙂

dhr reviewed (commented):
Thanks for the change! And very sorry I didn't take a closer look at this until recently. After testing internally, there are a couple of issues as detailed more thoroughly in the comment. Let me know if you have any thoughts on how to address those.

wprazuch commented:
@googlebot I fixed it.

wprazuch commented:
@googlebot I consent.

gmrukwa commented:
@sguada, @dhr, I fixed the issues you explained with the `__reduce__` approach 🙂 Perhaps we could merge it now, when you review it and I polish the details according to your suggestions. I've also added two tests that I skip by default, but it should be considered what should happen in these scenarios. I believe these are replicating what happens in Sonnet, though the behavior of the current code version against these tests is exactly the same as with the original code.

dhr commented on gin/config.py:
```diff
@@ -248,12 +248,17 @@ def _decorate_fn_or_cls(decorator, fn_or_cls, subclass=False):
       __doc__ = fn_or_cls.__doc__
       __module__ = fn_or_cls.__module__
-
+      __class__ = klass
```
I wonder if we could remove the "_gin_decorated" attribute and associated base class traversal above if we use `__class__ = fn_or_cls.__class__` here? I think that would propagate the right `__class__` down even when decorated multiple times (although haven't tested).

gmrukwa replied:
I've tested that (or at least what I think you think of) and it didn't work so smoothly for external configurables 🙂

dhr reviewed (commented):
Thanks! This all looks really good to me. And special thanks for the thorough test additions. I haven't had a chance to try it all out internally yet, will do that tomorrow and if that looks good will approve and merge it in.

gmrukwa commented:
@dhr, @sguada, can we merge it now that all conflicts got resolved?

gmrukwa commented:
@dhr, @sguada any news? :)

dhr commented:
Hey, sorry for the delay again. I think we're close to resolving everything... Unfortunately the approach with __reduce__ and overriding __class__ ran into a bunch of incompatibilities as well, involving some unforeseen errors raised by an internal library used in some projects.

Instead of changing that library (may be possible, but would take longer) I've prototyped a solution using wrapt to proxy the classes instead. There have quirks with that approach as well, but I think it's very close to working (still resolving one or two remaining issues). Once that's good to go/checked in (hopefully today), I'll merge most of your code here (in particular the added tests, which have been very helpful).

Thanks again for all the contributions and sorry for the slow turnaround!

dhr commented:
Further update... The wrapt approach also didn't pan out in the end, due to some fast-path type checks in NumPy (using lower-level CPython APIs) that don't play well with proxies. However, I had a new idea that finally seems to actually work and is backwards compatible with all internal projects. I think it's also the most principled/cleanest solution in the end. That's over in https://github.com/google/gin-config/pull/100. Once that's checked in I'll finally merge the tests here.

valentindey commented:
#100 has been merged but reverted again in 98531205efafcad4f41bd862d3d70b8ab8cda6ea . Are there any plans on how to proceed with this?

dhr commented:
This should finally all be resolved. There were some tricky to debug issues with the original commit #100 that required a rollback (in particular in the end it turned out there were some internal users who were dynamically subclassing registered/external configurable classes who were broken by the change). I finally had time to reinvestigate and fix things up... a roll-forward with those fixes just landed in #116, and the tests from this pull request were merged in https://github.com/google/gin-config/commit/26d134e0a40df87e7b14ed6665c0ee8382e802e2. Closing this PR now, sorry for the long delay!

gmrukwa commented:
Thanks, @dhr! It took a long time but was a pleasure to collaborate 🙂
